libraries/AP_HAL_ChibiOS/CANFDIface.cpp

```diff
@@ -647,10 +661,24 @@ bool CANIface::init(const uint32_t bitrate, const OperatingMode mode)
                   (timings.bs2 << FDCAN_NBTP_NTSEG2_Pos)  |
                   (timings.prescaler << FDCAN_NBTP_NBRP_Pos));
 
-    can_->DBTP = ((timings.bs1 << FDCAN_DBTP_DTSEG1_Pos) |
-                  (timings.bs2 << FDCAN_DBTP_DTSEG2_Pos)  |
-                  (timings.prescaler << FDCAN_DBTP_DBRP_Pos));
-    
+    if (fdbitrate) {
+        if (!computeTimings(fdbitrate, timings)) { // Do 8x fast Data transmission for CAN FD frames
+            can_->CCCR &= ~FDCAN_CCCR_INIT;
+            uint32_t while_start_ms = AP_HAL::millis();
+            while ((can_->CCCR & FDCAN_CCCR_INIT) == 1) {
+                if ((AP_HAL::millis() - while_start_ms) > REG_SET_TIMEOUT) {
```

Comment:
this is up to 250ms of interrupts disabled. Looks like we have this issue already in other places. Can it really take 250ms?
also, on boards with 16 bit timers, millis may stop advancing after 65ms as we've blocked the timer wrap interrupt, so we'd lockup I think
I think we need a helper function for this CCCR_INIT wait loop, and we need to arrange for it to not have interrupts disabled (I don't see why we'd need interrupts disabled while waiting for CCCR_INIT to change)

Reply:
It can take this long when there's an issue on the Transceiver hardware generating Bus Off condition. We do want to lock here I think, as Interrupts might not have been disabled yet. One of the interrupts is Bus Off Irq, which sets CCCR value, and could lead to race condition if interrupts are enabled before IE registers are set to 0.

Reply:
Maybe we can set initialised_ and not handle busoff error when initialised_ is false or rather have a semaphore. Lock is unnecessary here.

Reply:
I am creating a separate PR for this though. As you said this problem is much wider than just this location.